server/src/main/java/io/crate/execution/ddl/tables/TransportAddColumnAction.java

```diff
@@ -0,0 +1,314 @@
+                                       AddColumnRequest request,
+                                       ConvertedRequest convertedRequest) throws IOException {
+        Index[] concreteIndices = resolveIndices(currentState, request.relationName().indexNameOrAlias());
+
+        for (Index index : concreteIndices) {
+            final IndexMetadata indexMetadata = currentState.metadata().getIndexSafe(index);
+
+            Map<String, Object> indexMapping = indexMetadata.mapping().sourceAsMap();
+            mergeDeltaIntoExistingMapping(indexMapping, request, convertedRequest);
+            TransportSchemaUpdateAction.populateColumnPositions(indexMapping);
+
+            MapperService mapperService = indicesService.createIndexMapperService(indexMetadata);
+
+            mapperService.merge(indexMapping, MapperService.MergeReason.MAPPING_UPDATE);
+            DocumentMapper mapper = mapperService.documentMapper();
+
+            IndexMetadata.Builder imBuilder = IndexMetadata.builder(indexMetadata);
+            imBuilder.putMapping(new MappingMetadata(mapper.mappingSource())).mappingVersion(1 + imBuilder.mappingVersion());
+            metadataBuilder.put(imBuilder); // implicitly increments metadata version.
+        }
+
+        return ClusterState.builder(currentState).metadata(metadataBuilder).build();
+    }
+
+    private static void mergeDeltaIntoExistingMapping(Map<String, Object> existingMapping,
+                                                      AddColumnRequest request,
+                                                      ConvertedRequest convertedRequest) {
+        // _meta fields
+        if (convertedRequest.primaryKeys.isEmpty() == false) {
+            Maps.mergeInto(existingMapping, "_meta", List.of("primary_keys"), convertedRequest.primaryKeys,
+                (map, key, value) -> ((ArrayList<String>) map.computeIfAbsent(key, k -> new ArrayList<String>())).addAll(convertedRequest.primaryKeys)
+            );
+        }
+        if (convertedRequest.notNullColumns.isEmpty() == false) {
+            Maps.mergeInto(existingMapping, "_meta", List.of("constraints", "not_null"), convertedRequest.notNullColumns,
+                (map, key, value) -> {
+                    if ("not_null".equals(key)) {
+                        // we are merging on the end pf the path, i.e adding to the list
+                        ((ArrayList<String>) map.computeIfAbsent(key, k -> new ArrayList<String>())).addAll(convertedRequest.notNullColumns);
+                    } else {
+                        // When adding not-null for the first time,
+                        // Maps.mergeInto stops mid-path and transforms value (list) into chain of maps (to complement the remaining path) and adds value to the end.
+                        // There is no existing constraints in this case so we can take value (convertedRequest.notNullColumns) as is.
+                        map.put(key, value);
```

Comment:
other constraints don't need special handling - it's basically because of 
1. different "leaf" type vs what's produces by `nestedMaps` (List vs Map), so rules out gen/checks
2. long path, leading to situation where we stop mid way and instead of expected list get Map
I added integration tests locally and checks all 4 types on nested columns - since not null is the trickiest I think it's fine to have tests only for it?

Reply:
different now after https://github.com/crate/crate/commit/a835d593b20ecf58b3e21b16dcbdc5a56638f372, MergeInto is not used